src/patterns/ec2-app.ts

```diff
@@ -0,0 +1,105 @@
+import { GuVpc, SubnetType } from "../constructs/ec2";
+import { GuInstanceRole } from "../constructs/iam";
+import {
+  GuApplicationListener,
+  GuApplicationLoadBalancer,
+  GuApplicationTargetGroup,
+} from "../constructs/loadbalancing";
+
+interface GuEc2AppProps extends AppIdentity {
+  userData: GuUserData | string;
+  publicFacing: boolean; // could also name it `internetFacing` to match GuApplicationLoadBalancer
+  applicationPort: number;
+}
+
+interface GuMaybePortProps extends Omit<GuEc2AppProps, "applicationPort"> {
+  applicationPort?: number;
+}
+
+export enum GuApplicationPorts {
+  Node = 3000,
+  Play = 9000,
+}
+
+/*
+This pattern is under development. Please don't attempt to use it yet.
+ */
+export class GuEc2App {
+  constructor(scope: GuStack, props: GuEc2AppProps) {
+    const id = (id: string) => `${props.app}${id}`;
+
+    scope.tags.setTag(id("App"), props.app);
+
+    const { app } = props;
+    const vpc = GuVpc.fromIdParameter(scope, id("VPC"), { app });
+    const privateSubnets = GuVpc.subnetsfromParameter(scope, { type: SubnetType.PRIVATE, app });
+
+    const certificateArn = new GuArnParameter(scope, id("CertArn"), {
+      description: "ARN of a TLS certificate to install on the load balancer",
+    });
+
+    const certificate = Certificate.fromCertificateArn(scope, id("Certificate"), certificateArn.valueAsString);
+
+    const asg = new GuAutoScalingGroup(scope, id("AutoScalingGroup"), {
+      vpc,
+      app: props.app,
+      stageDependentProps: {
+        CODE: { minimumInstances: 1 },
+        PROD: { minimumInstances: 3 },
+      },
+      role: new GuInstanceRole(scope, { app: props.app }),
+      healthCheck: HealthCheck.elb({ grace: Duration.minutes(2) }), // should this be defaulted at pattern or construct level?
+      userData: props.userData instanceof GuUserData ? props.userData.userData : props.userData,
+      vpcSubnets: { subnets: privateSubnets },
+    });
+
+    const loadBalancer = new GuApplicationLoadBalancer(scope, id("LoadBalancer"), {
+      app,
+      vpc,
+      vpcSubnets: {
+        subnets: props.publicFacing ? GuVpc.subnetsfromParameter(scope, { type: SubnetType.PUBLIC }) : privateSubnets,
+      },
+    });
+
+    const targetGroup = new GuApplicationTargetGroup(scope, id("TargetGroup"), {
+      app,
+      vpc,
+      protocol: ApplicationProtocol.HTTP,
+      targets: [asg],
+      port: props.applicationPort,
+    });
+
+    new GuApplicationListener(scope, id("Listener"), {
+      app,
+      loadBalancer: loadBalancer,
+      defaultAction: ListenerAction.forward([targetGroup]),
+      certificates: [certificate],
+    });
+  }
+}
+
+/*
+This pattern is under development. Please don't attempt to use it yet.
+ */
+export class GuPlayApp extends GuEc2App {
+  constructor(scope: GuStack, props: GuMaybePortProps) {
+    super(scope, { ...props, applicationPort: props.applicationPort ?? GuApplicationPorts.Play });
```

Comment:
I'd favour being strongly opinionated here and say Play always runs on `9000`. That is, don't accept `applicationPort` in the props.

Reply:
Good shout, done.